### PR TITLE
Initial implementation of self damage

### PR DESCRIPTION
**Pull request by Charlie-Zheng**
Set up the initial implementation of self damage. For hyperbloom, burgeon, bloom, and burning

- [x] Remove HP
- [x] Blocked by shields
- [ ] Modified by resistance
- [ ] Can vape/melt/OL/apply pyro (for burning)

Player characters currently don't have resistance implemented, so it will not be done in this pull request.
Player self auras are also not fully supported, so the reactions for Burning will be omitted

### CODE REVIEW

**Comment by Charlie-Zheng** [comment]
[burgeon.txt](https://github.com/genshinsim/gcsim/files/12434785/burgeon.txt)
[P2W Nilou movement abuse.txt](https://github.com/genshinsim/gcsim/files/12434786/P2W.Nilou.movement.abuse.txt)
[temp.txt](https://github.com/genshinsim/gcsim/files/12434787/temp.txt)
[Baizhu burgeon.txt](https://github.com/genshinsim/gcsim/files/12434788/Baizhu.burgeon.txt)

The above are some config files I used to test the self damage implementation

File: pkg/avatar/avatar.go [hunk 1]
```diff
@@ -30,10 +33,199 @@ func (p *Player) Type() targets.TargettableType { return targets.TargettablePlay
...
+			Write("hp_per", atk.Snapshot.Stats[attributes.HPP]).
+			Write("total_hp", totalhp).
+			Write("catalyzed", atk.Info.Catalyzed).
+			Write("flat_dmg", atk.Info.FlatDmg).
+			Write("total_atk_def", a).
+			Write("base_dmg", base).
+			Write("ele", st).
+			Write("ele_per", elePer).
+			Write("bonus_dmg", dmgBonus).
+			Write("ignore_def", atk.Info.IgnoreDefPercent).
```
Comment (k0l11) [inline pkg/avatar/avatar.go:200]: Please add back the def_adj log right below this line, but set it to 0 with a comment saying that there are no DefMods applied to the player in sim.

File: pkg/avatar/avatar.go [hunk 2]
```diff
@@ -30,10 +33,199 @@ func (p *Player) Type() targets.TargettableType { return targets.TargettablePlay
...
+			ActorIndex: active,
+			Abil:       atk.Info.Abil,
+			Amount:     dmgLeft,
+		})
+	}
+	evt.Write("target", p.Key()).
+		Write("attack-tag", atk.Info.AttackTag).
+		Write("ele", atk.Info.Element.String()).
+		Write("damage", &dmgLeft).
+		Write("damage_before_shield", &dmg).
```
Comment (k0l11) [inline pkg/avatar/avatar.go:59]: Please swap these two lines (makes more sense to me to see dmg before shield first).

**Comment by k0l11** [comment]
Can you please add two new events [here](https://github.com/genshinsim/gcsim/blob/d0e14cdc4a237c006b2773a50c0b507879bbd0bd/pkg/core/player/shield/handler.go#L91C2-L91C2) so that users can see changes to shield hp in the shield debug category?
1. "shield taking dmg" should contain: 
- shield name
- shield element
- expiry
- dmg
- shield hp
- dmg after shield
- shield hp after dmg

3. "shield destroyed"
- shield name
- shield element
- expiry

**Review by k0l11 — APPROVED** [review]

**Pull request merged**